obstacle_distance: combine sensor distances to fit within distances array

rmackay9 opened the pull request:
This PR attempts to resolve issue https://github.com/mavlink/mavros/issues/1082.  The issue is that the obstacle_distance plugin does not correctly handle LaserScan messages from sensors with more than 72 distances.  The limitation comes from the [OBSTACLE_DISTANCE mavlink message](https://mavlink.io/en/messages/common.html#OBSTACLE_DISTANCE) which can only carry up to 72 distances.

The existing code simply truncates the data while the new code instead "compresses" the data to fit.  It does this by combining adjacent distances (always choosing the shortest distance).

So as to ensure as little impact as possible on existing users of this plugin, the new "compression" code is separated out with an if/else.

I've tested this on an Nvidia Tx2 running ROS/mavros with an RPLidarA2.  I modified the RPLidar's node.cpp file to send messages of three sizes: 36 distances, 72 distances and 360 distances.  The distances were hard-coded to always be index * 10cm to make it easy to check the results.

>    // testing code to send LaserScan to mavros::obstacle_distance plugin
>     scan_msg.intensities.resize(36);
>     scan_msg.ranges.resize(36);
>     scan_msg.angle_increment = 360 / 36 * M_PI / 180.0f;
>     for (size_t i = 0; i < scan_msg.ranges.size(); i++) {
>         scan_msg.ranges[i] = 0.1f * i;
>     }
>     scan_msg.range_max = 0.1f * scan_msg.ranges.size() + 1;
>     pub->publish(scan_msg);

Below are the testing results showing there are no changes to behaviour when there are 36 or 72 distances.
![before-36-and-72-slots](https://user-images.githubusercontent.com/1498098/51907963-b3207600-240b-11e9-9574-504a88ded432.png)


Below are the testing results when more than 72 distances are sent.  In this case 360 distances are sent
![before-after-with-360deg-sensor](https://user-images.githubusercontent.com/1498098/51907287-b0248600-2409-11e9-9996-71e504cde450.png)

Any and all feedback more than welcome!

TSC21 commented:
@jkflying @nicovanduijn @mrivi @baumanta your reviews are welcomed.

vooon commented on mavros_extras/src/plugins/obstacle_distance.cpp:
```diff
@@ -64,15 +64,32 @@ class ObstacleDistancePlugin : public plugin::PluginBase {
-		Eigen::Map<Eigen::Matrix<uint16_t, Eigen::Dynamic, 1> > map_distances(obstacle.distances.data(), n);
+		if (req->ranges.size() <= obstacle.distances.size()) {
+			// all distances from sensor will fit in obstacle distance message
+			auto n = std::min(req->ranges.size(), obstacle.distances.size());
```
Not needed anymore, req->ranges.size() will be always less than obstacle.arr.

rmackay9 replied:
OK, removed "n" and substituted, "req->ranges.size()" wherever it was used.  To some extent I wanted to leave the existing code as it was to reduce worry but that's all good if you guys are happy.

vooon commented on mavros_extras/src/plugins/obstacle_distance.cpp:
```diff
@@ -64,15 +64,32 @@ class ObstacleDistancePlugin : public plugin::PluginBase {
+			obstacle.increment = req->angle_increment * RAD_TO_DEG;				//!< [degrees]
+		} else {
+			// all distances from sensor will not fit so we combine adjacent distances always taking the shortest distance
+			int scale_factor = ceil(req->ranges.size() / obstacle.distances.size());
```
It is Ok to use int for scale_factor (but i'd more like to see size_t). But one of sizes should be casted to FP before ceil.

```cpp
ceil(double(req->ranges.size()) / ...)
```

rmackay9 replied:
Txs very much for the reviews.  in a follow up commit:

- scale_factor, i and j are all changed to size_t
- ceil() argument is cast to double

vooon commented on mavros_extras/src/plugins/obstacle_distance.cpp:
```diff
@@ -64,15 +64,32 @@ class ObstacleDistancePlugin : public plugin::PluginBase {
+				for (int j = 0; j < scale_factor; j++) {
+					int req_index = i * scale_factor + j;
+					if (req_index < req->ranges.size()) {
+						uint16_t dist_cm = req->ranges[req_index] * 1e2;
```
That is not a problem.

vooon commented on mavros_extras/src/plugins/obstacle_distance.cpp:
```diff
@@ -64,15 +64,32 @@ class ObstacleDistancePlugin : public plugin::PluginBase {
+					if (req_index < req->ranges.size()) {
+						uint16_t dist_cm = req->ranges[req_index] * 1e2;
+						obstacle.distances[i] = std::min(obstacle.distances[i], dist_cm);
+					}
```
I have a bad idea to use `VectorXf::minCoef()` via Map...

rmackay9 replied:
@vooon, cool, I looked around a bit in Eigen to see if there was something to accomplish the same thing but I'm afraid I couldn't find anything.

hamishwillee commented:
Do we need to revisit [OBSTACLE_DISTANCE](https://mavlink.io/en/messages/common.html#OBSTACLE_DISTANCE) message?

rmackay9 commented on mavros_extras/src/plugins/obstacle_distance.cpp:
```diff
@@ -64,15 +64,32 @@ class ObstacleDistancePlugin : public plugin::PluginBase {
+			auto n = std::min(req->ranges.size(), obstacle.distances.size());
+			Eigen::Map<Eigen::Matrix<uint16_t, Eigen::Dynamic, 1> > map_distances(obstacle.distances.data(), n);
+			auto cm_ranges = Eigen::Map<const Eigen::VectorXf>(req->ranges.data(), n) * 1e2;
+			map_distances = cm_ranges.cast<uint16_t>();							//!< [centimeters]
```
@jkflying, txs for the review.  sure, aligned comments in follow up commit

rmackay9 commented:
I've added two commits which I think addresses the issues raised.  I've retested and confirmed there's no change to the final results.

As a side note, I didn't want to confuse things but I think that the existing calcs for the "increment" (the number of degrees covered by each element in the distances array) is often incorrect by "1" due the putting a floating point number into an integer field.

>> obstacle.increment = req->angle_increment * RAD_TO_DEG;				//!< [degrees]

This is an existing problem so I didn't correct it as part of this PR but I could address it in a follow-up PR if people like.  Basically I think we should use ceil() like in the new code.

rmackay9 commented:
In case it helps, I'm very happy to help maintain this little contribution going forward in case problems appear (which I don't think they will).

nicovanduijn commented:
Did someone test whether this breaks the collision prevention capabilites of PX4/avoidance?

rmackay9 commented:
@nicovanduijn, i'd be surprised if someone had done that testing (I'm not in a position to do that) but I have proven that the output before and after this change is the same for cases where the source sensor produces less than 72 slots.  I cannot imagine that existing consumers use more than 72 slots or they would surely have noticed the error this PR fixes.

nicovanduijn commented:
Good enough for me. The current avoidance is broken for the most recent mavros anyway (Fix ready in https://github.com/PX4/avoidance/pull/197)
